File: GameStudies/CS/Personagem.cs
using System.Collections.Generic;
using UnityEngine;

public class Personagem : MonoBehaviour
{
    public string Nome;

    public int Vida = 100;
    public int Energia = 50;

    public float velocidade = 5f;

    private List<string> dialogos = new List<string>();

    private List<string> missoes = new List<string>();

    void Update()
    {
        Movimentar();
    }

    void Movimentar()
    {
        float dx = Input.GetAxis("Horizontal");
        float dy = Input.GetAxis("Vertical");

        Vector2 movimento = new Vector2(dx, dy);
        transform.Translate(movimento * velocidade * Time.deltaTime);
    }

    public void AdicionarDialogo(string fala)
    {
        dialogos.Add(fala);
    }

    public void Falar(int index)
    {
        if (index >= 0 && index < dialogos.Count)
        {
            Debug.Log(Nome + ": " + dialogos[index]);
        }
    }

    public void AdicionarMissao(string missao)
    {
        missoes.Add(missao);
    }

    public void ListarMissoes()
    {
        foreach (var m in missoes)
        {
            Debug.Log(m);
        }
    }

    public void Atacar(Personagem alvo)
    {
        int dano = 10;
        alvo.TomarDano(dano);
    }

    public void TomarDano(int dano)
    {
        Vida -= dano;
        if (Vida < 0) Vida = 0;
    }
}

File: GameStudies/CPP/Personagem.cpp
#include "Personagem.hpp"
#include <iostream>

Personagem::Personagem(std::string nome) {
    this->nome = nome;
    vida = 100;
    x = 0;
    y = 0;
}

void Personagem::mover(float dx, float dy) {
    x += dx;
    y += dy;
}

void Personagem::tomarDano(int dano) {
    vida -= dano;
    if (vida < 0) vida = 0;
}

std::string Personagem::getNome() {
    return nome;
}